docs/docs/guides/js_apps/authwit.md

```diff
@@ -0,0 +1,117 @@
+:::note
+
+If you are using public authwit (ie using `assert_current_call_valid_authwit_public` in your contract), you will need to deploy the following accounts publicly:
+
+1. The account that is giving permission to an account to act on behalf of it (authwit giver)
+2. The account that does the action (authwit receiver)
+
+Here is an example implementation:
+
+#include_code public_deploy_accounts yarn-project/end-to-end/src/fixtures/utils.ts typescript
+
+You would then call this like so:
+
+#include_code public_deploy_accounts yarn-project/end-to-end/src/e2e_authwit.test.ts typescript
+
+## Define the action
+
+When creating an authwit, you will need to pass the authwit givier, the authwit receiver (who will perform the action), and the action that is being authorized. The action can be a smart contract function call, or alternatively arbitrary data.
+
+### When the action is a function call
+
+You can define the action like this:
+
+#include_code authwit_computeAuthWitMessageHash yarn-project/end-to-end/src/e2e_blacklist_token_contract/transfer_private.test.ts typescript
+
+In this example,
+
+- `asset` refers to a token contract
+- `withWallet(wallets[1])` is specifying the authwit receiver (`wallets[1]`) will do this action
+- `.methods.transfer()` is specifying that the action is calling the `transfer` method on the token contract
+- `(wallets[0].getAddress(), wallets[1].getAddress(), amount, nonce);` are the args of this method - it will send the `amount` from `wallets[0]` to `wallets[1]`
+
+### Arbitrary message
+
+You can hash your own authwit message by creating an inner hash with the data, like this:
+
+#include_code compute_inner_authwit_hash yarn-project/end-to-end/src/e2e_authwit.test.ts typescript
+
+Then create the outer hash by hashing the inner hash with the authwit receiver address, chainId, and version:
+
+#include_code compute_outer_authwit_hash yarn-project/end-to-end/src/e2e_authwit.test.ts typescript
+
+## Create the authwit
+
+These are slightly different interfaces depending on whether your contract is checking the authwit in private or public. As public authwits are stored in the account contract and batched with the authwit action call, it is done with one transaction. Private execution uses oracles, so the authwit needs to be created by the authwit giver and then added to the receiver's wallet.
+
+### Private
+
+This is expected to be used alongside [private authwits in Aztec.nr contract](../smart_contracts/writing_contracts/authwit.md#private-functions).
+
+Create a private authwit like this:
+
+#include_code create_authwit yarn-project/end-to-end/src/e2e_blacklist_token_contract/transfer_private.test.ts typescript
+
+In this example,
+
+- `wallets[0]` is the authwit giver
+- `wallets[1]` is the authwit reciever and caller of the function
+- `action` was [defined previously](#define-the-action)
+
+If you created an artbitrary message, you can create the authwit by replacing these params with the outer hash:
```

Comment:
Just curious what some use cases of creating authwits for arbitrary messages would be. Should we add some examples here or in another doc?

Reply:
i can't think of any myself. it's a comment in the codebase here https://github.com/AztecProtocol/aztec-packages/blob/a265da0a0e4cb5666c91812dc725f5620ad9b740/yarn-project/aztec.js/src/utils/authwit.ts#L38
have asked jan